lib/nmatrix/shortcuts.rb

```diff
@@ -275,6 +275,61 @@ def diagonal(entries, opts={})
     alias :diag :diagonal
     alias :diagonals :diagonal
 
+    # Generate a block-diagonal NMatrix from the supplied 2D square matrices.
+    #
+    # * *Arguments*
+    #   - +*params+ -> An array that collects all arguments passed to the method. The method
+    #                  can receive any number of arguments. Optionally, the last entry of +params+ is 
+    #                  a hash of options from NMatrix#initialize. All other entries of +params+ are 
+    #                  the blocks of the desired block-diagonal matrix, which are supplied
+    #                  as square 2D NMatrix objects.
+    # * *Returns*
+    #   - NMatrix of block-diagonal form filled with specified matrices
+    #     as the blocks along the diagonal.
+    #
+    # * *Example* 
+    #
+    #  a = NMatrix.new([2,2],[1,2,3,4])
+    #  b = NMatrix.new([1,1],[123],dtype: :int32)
+    #  c = NMatrix.new([3,3],[1,2,3,1,2,3,1,2,3], dtype: :float64)
+    #  m = NMatrix.block_diagonal(a,b,c,dtype: :int64, stype: :yale)
+    #      => 
+    #      [
+    #        [1, 2,   0, 0, 0, 0]
+    #        [3, 4,   0, 0, 0, 0]
+    #        [0, 0, 123, 0, 0, 0]
+    #        [0, 0,   0, 1, 2, 3]
+    #        [0, 0,   0, 1, 2, 3]
+    #        [0, 0,   0, 1, 2, 3]
+    #      ]
+    #
+    def block_diagonal(*params)
+      options = params.last.is_a?(Hash) ? params.pop : {}
+
+      block_sizes = [] #holds the size of each matrix block
+      params.each do |b|
+        raise ArgumentError, "Only NMatrix objects allowed" unless b.is_a?(NMatrix)
```

Comment:
I think it is a good idea to support Array inputs as well, though I think that it's more intuitive when the input is a matrix.
Thinking about it now, I also find that it would be good to have the possibility to supply an array of NMatrix objects. Assume, I want a block-diagonal matrix that has another matrix `a` 100 times along its diagonal. Clearly, I wouldn't want to supply 100 arguments `a` to the block-diagonal method. So, I think that the method should also support something like:

```Ruby
a = NMatrix.ones([2,2])
a100 = Array.new(100) { |i| a }
m = NMatrix.block_diagonal(a100)
```

What do you think?
The only way, I can think of right now, of how to incorporate all of that, is to have a bunch of `if` statement in the code, which would make it bloated. Is there possibly a more elegant way?